### DISCUSSION TITLE
Why is the ws adapter using JSON.parse and JSON.stringify?

### DISCUSSION THREAD

**Discussion by gunhaxxor** [trpc/trpc#3969] [Q&A]
I was looking into the code of the ws adapter.
In this section the adapter is using `JSON.parse` to deserialize the incoming payload:
https://github.com/trpc/trpc/blob/783f9bcae31ef788962a4417bda15f190598ecb7/packages/server/src/adapters/ws.ts#L276-L283
This is happening before the transformer receives the message(s). 

And here the adapter uses `JSON.stringify` when sending:
https://github.com/trpc/trpc/blob/783f9bcae31ef788962a4417bda15f190598ecb7/packages/server/src/adapters/ws.ts#L122-L126

I thought the idea of the transformers was to be able to use different implementations of serialization/deserialization. Doesn't this use of JSON.parse and JSON.stringify defeat that purpose?

I'm trying to use [msgpackr](https://www.npmjs.com/package/msgpackr) as a transformer and I think this is what makes my setup not work at the moment. Perhaps someone with insight into the websocket part of the repo might be able to help?
@sachinraja I got the impression you have great insight into the ws-adapter, perhaps you know?

Grateful for any suggenstions/clarifications!

**Comment by kamilogorek** [comment]
note: There's some discussion about similar case in https://github.com/trpc/trpc/pull/3804 and referenced issue.

**Comment by gunhaxxor** [comment]
Ok. My immediate assumption was that the transformers would be used on the "whole" payload
```ts
{
  id,
  jsonrpc,
  result: {
    type: 'data',
    data: result,
  },
}
```
But both the normal http stuff and the ws-adapter seems to do JSON.parse and stringify for the whole payload and *then* apply the transformer to the result part. I'm not sure whats the purpose of that? Wouldn't it be more lean to apply the transformer to the whole payload? At least on the websocket side I can't see why that would be bad 🤔

Is it perhaps related to the input needing to be compatible with url query-strings for the http stuff?


In my use case I've pinpointed the problem to be that my transformer's serializer outputs Buffer/Uint8Array and the deserializer naturally expects that as an input as well. But JSON.stringify of a Uint8Array is not preserved when using JSON.parse.
So the flow becomes.
when sending:
js-data -> Uint8Array -> string that looks like a json-object (a dictionary with array index as keys)
when receiving:
string -> Object -> not deserializable by my transformer...
